phd-tests/framework/src/disk/crucible.rs

```diff
@@ -0,0 +1,275 @@
+    ) -> anyhow::Result<Self> {
+        // To create a region, Crucible requires a block size, an extent size
+        // given as a number of blocks, and an extent count. Compute the latter
+        // two figures here. The extent size is semi-arbitrarily chosen to be 64
+        // MiB (to match Omicron's extent size at the time this module was
+        // authored); this can be parameterized later if needed.
+        const EXTENT_SIZE: u64 = 64 << 20;
+
+        assert!(EXTENT_SIZE > block_size.bytes());
+        assert!(EXTENT_SIZE % block_size.bytes() == 0);
+
+        let blocks_per_extent: u64 = EXTENT_SIZE / block_size.bytes();
+        let extents_in_disk = (disk_size_gib * (1 << 30)) / EXTENT_SIZE;
+
+        // Create the region directories for each region.
+        let mut data_dirs = vec![];
+        let disk_uuid = Uuid::new_v4();
+        for port in downstairs_ports {
+            let mut data_dir_path = data_dir.as_ref().to_path_buf();
+            data_dir_path.push(format!("{}_{}", disk_uuid, port));
+            std::fs::create_dir_all(&data_dir_path)?;
+            data_dirs.push(DataDirectory { path: data_dir_path });
+        }
+
+        for dir in &data_dirs {
+            let dir_arg = dir.path.to_string_lossy();
+            let crucible_args = [
+                "create",
+                "--block-size",
+                &block_size.bytes().to_string(),
+                "--data",
+                dir_arg.as_ref(),
+                "--encrypted",
+                "false",
+                "--uuid",
+                &disk_uuid.to_string(),
+                "--extent-size",
+                &blocks_per_extent.to_string(),
+                "--extent-count",
+                &extents_in_disk.to_string(),
+            ];
+
+            // This is a transient process, so just pipe stdout/stderr back into
+            // the framework and trace the outputs instead of setting up full
+            // log files.
+            let create_stdout = Stdio::piped();
+            let create_stderr = Stdio::piped();
+            let create_proc = run_crucible_downstairs(
+                &downstairs_binary_path,
+                &crucible_args,
+                create_stdout,
+                create_stderr,
+            )?;
+
+            let create_output = create_proc.wait_with_output()?;
+            info!(
+                stdout = std::str::from_utf8(&create_output.stdout)?,
+                stderr = std::str::from_utf8(&create_output.stderr)?,
+                status = ?create_output.status,
+                "Created Crucible region using crucible-downstairs"
+            );
+
+            if !create_output.status.success() {
+                anyhow::bail!(
+                    "Crucible region creation failed with exit code {:?}",
+                    create_output.status.code()
+                );
+            }
+        }
+
+        // Spawn the downstairs processes that will serve requests from guest
```

Comment:
I don't want to encourage you to duplicate all of dsc here, but you might want the ability to create a region set, then come back later and start the downstairs on that already created region set.  It may not be something you care about, but it can be faster sometime to re-use an existing region vs. creating from scratch each time.

Reply:
I think this would be interesting for tests that verify that Propolis can be made to do the right thing if a downstairs disappears out from under a running instance and is subsequently replaced (we would want the test to emulate this by terminating the downstairs, starting a new one, and then calling whatever sled agent/Nexus would call to advise Propolis of the new downstairs address). I might defer this logic until we're ready to try that out, though.

Reply:
Yeah, I just wanted you to be aware of what dsc has to offer :)